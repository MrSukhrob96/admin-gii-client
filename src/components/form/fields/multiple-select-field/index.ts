export * from './multiple-select-field.component';
