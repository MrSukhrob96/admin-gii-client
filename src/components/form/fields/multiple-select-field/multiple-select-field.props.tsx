import { Button, MenuItem, OptionProps } from '@blueprintjs/core';
import { ItemRenderer } from '@blueprintjs/select';
import { useEffect, useMemo, useState } from 'react';
import { Control, ControllerProps, useController } from 'react-hook-form';

/**
 * <MultipleSelectField /> props
 */
export type MultipleSelectFieldProps = {
  items: OptionProps[];
  name: string;
  helperText?: string;
  label?: string;
  rules?: ControllerProps['rules'];
  control: Control<any>;
};

export const useMultipleSelectFieldProps = (
  props?: MultipleSelectFieldProps
) => {
  const [selectedItems, setSelectedItems] = useState([]);

  const {
    field: { value, onChange, onBlur },
    fieldState: { error }
  } = useController({
    name: props.name,
    rules: props.rules,
    control: props.control
  });

  const onItemItemSelect = (item: OptionProps) => {
    const isSelected = findSelectedItem(item);
    let newItems;
    if (isSelected) {
      newItems = selectedItems.filter(
        selectedItem => selectedItem.value !== item.value
      );
    } else {
      newItems = [...selectedItems, item];
    }
    onChange(newItems.map(selectedItem => selectedItem.value));
  };

  const itemTagRenderer = (item: OptionProps) => item.label;

  const findSelectedItem = (item: OptionProps) => {
    return selectedItems.find(
      selectedItem => selectedItem.value === item.value
    );
  };

  const handleItemRemove = (label: string, index: number) => {
    onChange(selectedItems.filter((_, i) => i !== index).map(selectedItem => selectedItem.value));
  };

  const itemRenderer: ItemRenderer<OptionProps> = (
    item: OptionProps,
    { modifiers, index, handleClick }
  ) => {
    if (!modifiers.matchesPredicate) {
      return null;
    }
    return (
      <MenuItem
        selected={modifiers.active}
        icon={findSelectedItem(item) ? 'tick' : 'blank'}
        key={item.value}
        onClick={handleClick}
        text={`${index}. ${item.label}`}
        shouldDismissPopover={false}
      />
    );
  };
  const itemClearButton = useMemo(
    () =>
      selectedItems.length > 0 ? (
        <Button
          icon='cross'
          minimal={true}
          onClick={() => setSelectedItems([])}
        />
      ) : undefined,
    [selectedItems]
  );

  useEffect(() => {
    if (props.items?.length <= 0) return

    const items = value.map((val: string) => props.items.find(item => item.value === val))
    setSelectedItems(items);
  }, [props.items, value])

  return {
    itemTagRenderer,
    itemRenderer,
    onItemItemSelect,
    selectedItems,
    itemClearButton,
    handleItemRemove,
    error
  };
};
