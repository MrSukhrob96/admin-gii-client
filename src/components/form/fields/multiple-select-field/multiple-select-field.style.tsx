import styled from 'styled-components'

export const MultipleSelectFieldContainer = styled.div`

`