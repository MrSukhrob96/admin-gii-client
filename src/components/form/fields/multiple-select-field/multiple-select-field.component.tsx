import { FormGroup, MenuItem } from '@blueprintjs/core';
import { MultiSelect } from '@blueprintjs/select';
import { hoc } from '@utils/hoc';
import { useMultipleSelectFieldProps } from './multiple-select-field.props';

/**
 * <MultipleSelectField />
 */
export const MultipleSelectField = hoc(
  useMultipleSelectFieldProps,
  ({
    items,
    itemTagRenderer,
    itemRenderer,
    onItemItemSelect,
    itemClearButton,
    handleItemRemove,
    name,
    label,
    rules,
    helperText,
    error,
    selectedItems
  }) => (
    <FormGroup
      helperText={error?.message || helperText}
      label={label}
      labelFor={`${name}-input`}
      labelInfo={rules?.required && '(majburiy)'}
      intent={error ? 'danger' : 'none'}
    >
      <MultiSelect
        items={items}
        tagRenderer={itemTagRenderer}
        itemRenderer={itemRenderer}
        onItemSelect={onItemItemSelect}
        selectedItems={selectedItems || []}
        noResults={<MenuItem disabled={true} text='No results.' />}
        tagInputProps={{
          onRemove: handleItemRemove,
          rightElement: itemClearButton
        }}
        fill
      />
    </FormGroup>
  )
);
