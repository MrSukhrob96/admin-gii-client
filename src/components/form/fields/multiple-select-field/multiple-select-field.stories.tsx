import { ComponentStory, ComponentMeta } from '@storybook/react';
import { MultipleSelectField } from './multiple-select-field.component';
import '@styles/index.css';

export default {
  title: 'Common/MultipleSelectField',
  component: MultipleSelectField,
} as ComponentMeta<typeof MultipleSelectField>

const Template: ComponentStory<typeof MultipleSelectField> = (args) => <MultipleSelectField {...args} />;

export const MultipleSelectFieldInitial = Template.bind({});
MultipleSelectFieldInitial.args = {
}