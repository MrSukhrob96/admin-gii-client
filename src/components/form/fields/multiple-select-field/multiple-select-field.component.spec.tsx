import { customRender } from '@test-utils';
import { renderHook } from '@testing-library/react-hooks';
import { useForm } from 'react-hook-form';
import { MultipleSelectField } from './multiple-select-field.component';

describe('MultipleSelectField', () => {
  test('Should render MultipleSelectField component', async () => {
    const { result } = renderHook(() => useForm({
      defaultValues: { test: '' }
    }));
    const { asFragment } = customRender(
      <MultipleSelectField control={result.current.control} name='test' items={[]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});