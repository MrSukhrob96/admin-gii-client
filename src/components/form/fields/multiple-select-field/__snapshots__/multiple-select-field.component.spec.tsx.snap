// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleSelectField Should render MultipleSelectField component 1`] = `
<DocumentFragment>
  <div
    class="bp4-form-group"
  >
    <div
      class="bp4-form-content"
    >
      <div
        class="bp4-popover-wrapper bp4-fill"
      >
        <div
          aria-haspopup="true"
          class="bp4-popover-target"
        >
          <div
            class=""
          >
            <div
              class="bp4-input bp4-tag-input bp4-fill bp4-multi-select"
            >
              <div
                class="bp4-tag-input-values"
              >
                <input
                  class="bp4-input-ghost bp4-multi-select-tag-input-input"
                  placeholder="Search..."
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
